File: ios/Classes/Location/LocationClient.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//

import Foundation
import CoreLocation

class LocationClient : NSObject, CLLocationManagerDelegate {
  
  private let locationManager = CLLocationManager()
  private var permissionCallbacks: Array<Callback<Void, Void>> = []
  private var permissionSettingsCallback: (() -> Void)? = nil
  
  private var locationUpdatesCallback: LocationUpdatesCallback? = nil
  private var locationUpdatesRequests: Array<LocationUpdatesRequest> = []
  
  private var hasLocationRequest: Bool {
    return !locationUpdatesRequests.isEmpty
  }
  private var hasInBackgroundLocationRequest: Bool {
    return !locationUpdatesRequests.filter { $0.inBackground == true }.isEmpty
  }
  
  private var isPaused = false
  
  override init() {
    super.init()
    locationManager.delegate = self
  }
  
  
  // One shot API
  
  func isLocationOperational(with permission: Permission) -> Result<Bool> {
    let status: ServiceStatus<Bool> = currentServiceStatus(with: permission)
    return status.isReady ? Result<Bool>.success(with: true) : status.failure!
  }
  
  func requestLocationPermission(with permission: PermissionRequest, _ callback: @escaping (Result<Bool>) -> Void) {
    runWithValidServiceStatus(with: permission, success: {
      callback(Result<Bool>.success(with: true))
    }, failure: { result in
      callback(result)
    })
  }
  
  func lastKnownLocation(with permission: Permission, _ callback: @escaping (Result<[Location]>) -> Void) {
    runWithValidServiceStatus(with: permission, success: {
      if let location = self.locationManager.location {
        callback(Result<Location>.success(with: [Location(from: location)]))
      } else {
        callback(Result<Location>.failure(of: .locationNotFound))
      }
    }, failure: callback)
  }
  
  
  // Updates API
  
  func addLocationUpdates(request: LocationUpdatesRequest) {
    runWithValidServiceStatus(with: request.permission, success: {
      self.locationUpdatesRequests.append(request)
      self.updateRunningRequest()
    }, failure: { result in
      self.locationUpdatesCallback?(result)
    })
  }
  
  func removeLocationUpdates(requestId: Int) {
    guard let index = locationUpdatesRequests.firstIndex(where: { $0.id == requestId }) else {
      return
    }
    
    locationUpdatesRequests.remove(at: index)
    updateRunningRequest()
  }
  
  func registerLocationUpdates(callback: @escaping LocationUpdatesCallback) {
    precondition(locationUpdatesCallback == nil, "trying to register a 2nd location updates callback")
    locationUpdatesCallback = callback
  }
  
  func deregisterLocationUpdatesCallback() {
    precondition(locationUpdatesCallback != nil, "trying to deregister a non-existent location updates callback")
    locationUpdatesCallback = nil
  }
  
  
  // Lifecycle API
  
  func resume() {
    if let callback = permissionSettingsCallback {
      callback()
      permissionSettingsCallback = nil
    }
    
    guard hasLocationRequest && isPaused else {
      return
    }
    
    isPaused = false
    startLocation()
  }
  
  func pause() {
    guard hasLocationRequest && !isPaused && !hasInBackgroundLocationRequest else {
      return
    }
    
    isPaused = true
    locationManager.stopUpdatingLocation()
  }
  
  
  // Location updates logic
  
  private func updateRunningRequest() {
    guard !locationUpdatesRequests.isEmpty else {
      isPaused = false
      locationManager.stopUpdatingLocation()
      return
    }
    
    locationManager.desiredAccuracy = locationUpdatesRequests.max(by: {
      let best = LocationHelper.betterAccuracy(between: $0.accuracy.clValue, and: $1.accuracy.clValue)
      return best == $0.accuracy.clValue
    })!.accuracy.clValue
    
    let distanceFilter = locationUpdatesRequests.map { $0.displacementFilter }.min()!
    locationManager.distanceFilter = distanceFilter > 0 ? distanceFilter : kCLDistanceFilterNone
    
    locationManager.stopUpdatingLocation()
    
    if !isPaused {
      startLocation()
    }
  }
  
  private func startLocation() {
    let isContinuousUpdatesRequested = !self.locationUpdatesRequests.filter { $0.strategy == .continuous }.isEmpty
    
    if isContinuousUpdatesRequested {
      locationManager.startUpdatingLocation()
    } else {
      locationManager.requestLocation()
    }
  }
  
  
  // Service status
  
  private func runWithValidServiceStatus<T>(with permission: Permission, success: @escaping () -> Void, failure: @escaping (Result<T>) -> Void) {
    let permissionRequest = PermissionRequest(value: permission, openSettingsIfDenied: false)
    runWithValidServiceStatus(with: permissionRequest, success: success, failure: failure)
  }
  
  private func runWithValidServiceStatus<T>(with permission: PermissionRequest, success: @escaping () -> Void, failure: @escaping (Result<T>) -> Void) {
    let status: ServiceStatus<T> = currentServiceStatus(with: permission.value)
    
    if status.isReady {
      success()
    } else {
      if let permission = status.needsAuthorization {
        let callback = Callback<Void, Void>(
          success: { _ in success() },
          failure: { _ in failure(Result<T>.failure(of: .permissionDenied)) }
        )
        permissionCallbacks.append(callback)
        locationManager.requestAuthorization(for: permission)
      } else {
        if #available(iOS 10.0, *),
          status.failure!.error!.type == .permissionDenied,
          permission.openSettingsIfDenied,
          let appSettingURl = URL(string: UIApplication.openSettingsURLString),
          UIApplication.shared.canOpenURL(appSettingURl) {
          permissionSettingsCallback = {
            let refreshedStatus: ServiceStatus<T> = self.currentServiceStatus(with: permission.value)
            if refreshedStatus.isReady {
              success()
            } else {
              failure(refreshedStatus.failure!)
            }
          }
          UIApplication.shared.openURL(appSettingURl)
        } else {
          failure(status.failure!)
        }
      }
    }
  }
  
  private func currentServiceStatus<T>(with permission: Permission) -> ServiceStatus<T> {
    guard CLLocationManager.locationServicesEnabled() else {
      return ServiceStatus<T>(isReady: false, needsAuthorization: nil, failure: Result<T>.failure(of: .serviceDisabled))
    }
    
    switch CLLocationManager.authorizationStatus() {
    case .notDetermined:
      guard locationManager.isPermissionDeclared(for: permission) else {
        return ServiceStatus<T>(isReady: false, needsAuthorization: nil, failure: Result<T>.failure(of: .runtime, message: "Missing location usage description values in Info.plist. See readme for details.", fatal: true))
      }
      
      return ServiceStatus<T>(isReady: false, needsAuthorization: permission, failure: Result<T>.failure(of: .permissionNotGranted))
    case .denied:
      return ServiceStatus<T>(isReady: false, needsAuthorization: nil, failure: Result<T>.failure(of: .permissionDenied))
    case .restricted:
      return ServiceStatus<T>(isReady: false, needsAuthorization: nil, failure: Result<T>.failure(of: .serviceDisabled))
    case .authorizedWhenInUse, .authorizedAlways:
      return ServiceStatus<T>(isReady: true, needsAuthorization: nil, failure: nil)
    @unknown default:
      fatalError("Unknown CLLocationManager.authorizationStatus(): \(CLLocationManager.authorizationStatus())")
    }
  }
  
  
  // CLLocationManagerDelegate
  
  public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
    permissionCallbacks.forEach { action in
      if status == .authorizedAlways || status == .authorizedWhenInUse {
        action.success(())
      } else {
        action.failure(())
      }
    }
    permissionCallbacks.removeAll()
  }
  
  public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    locationUpdatesCallback?(Result<[Location]>.success(with: locations.map { Location(from: $0) }))
  }
  
  public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    locationUpdatesCallback?(Result<[Location]>.failure(of: .runtime, message: error.localizedDescription))
  }
  
  struct Callback<T, E> {
    let success: (T) -> Void
    let failure: (E) -> Void
  }
  
  typealias LocationUpdatesCallback = (Result<[Location]>) -> Void
  
  struct ServiceStatus<T: Codable> {
    let isReady: Bool
    let needsAuthorization: Permission?
    let failure: Result<T>?
  }
}

File: android/src/main/kotlin/app/loup/geolocation/location/LocationClient.kt
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

package app.loup.geolocation.location

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.net.Uri
import android.provider.Settings
import androidx.core.app.ActivityCompat
import app.loup.geolocation.GeolocationPlugin
import app.loup.geolocation.data.*
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.*
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine


@SuppressLint("MissingPermission")
class LocationClient(private val context: Context) {

  var activity: Activity? = null

  val permissionResultListener: PluginRegistry.RequestPermissionsResultListener = PluginRegistry.RequestPermissionsResultListener { id, _, grantResults ->
    if (id == GeolocationPlugin.Intents.LocationPermissionRequestId) {
      if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
        permissionCallbacks.forEach { it.success(Unit) }
      } else {
        permissionCallbacks.forEach { it.failure(Unit) }
      }
      permissionCallbacks.clear()
      return@RequestPermissionsResultListener true
    }

    return@RequestPermissionsResultListener false
  }

  val activityResultListener: PluginRegistry.ActivityResultListener = PluginRegistry.ActivityResultListener { id, resultCode, _ ->
    when (id) {
      GeolocationPlugin.Intents.LocationPermissionSettingsRequestId -> {
        permissionSettingsCallback?.invoke()
        permissionSettingsCallback = null
        return@ActivityResultListener true
      }

      GeolocationPlugin.Intents.EnableLocationSettingsRequestId -> {
        if (resultCode == Activity.RESULT_OK) {
          locationSettingsCallbacks.forEach { it.success(Unit) }
        } else {
          locationSettingsCallbacks.forEach { it.failure(Unit) }
        }
        locationSettingsCallbacks.clear()

        return@ActivityResultListener true
      }
    }

    return@ActivityResultListener false
  }

  private val providerClient: FusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
  private val permissionCallbacks = ArrayList<Callback<Unit, Unit>>()
  private var permissionSettingsCallback: (() -> Unit)? = null
  private val locationSettingsCallbacks = ArrayList<Callback<Unit, Unit>>()

  private var locationUpdatesCallback: ((Result) -> Unit)? = null
  private val locationUpdatesRequests = ArrayList<LocationUpdatesRequest>()
  private var currentLocationRequest: LocationRequest? = null

  private val hasLocationRequest get() = currentLocationRequest != null
  private val hasInBackgroundLocationRequest get() = locationUpdatesRequests.any { it.inBackground }

  private var isPaused = false

  private val locationCallback: LocationCallback = object : LocationCallback() {
    override fun onLocationResult(result: LocationResult) {
      onLocationUpdatesResult(Result.success(result.locations.map { LocationData.from(it) }))
    }
  }

  suspend fun enableLocationServices(): Result {
    return if (LocationHelper.isLocationEnabled(context)) {
      Result.success(true)
    } else {
      Result(requestEnablingLocation())
    }
  }

  // One shot API

  fun isLocationOperational(permission: Permission): Result {
    val status = currentServiceStatus(permission)
    return if (status.isReady) {
      Result.success(true)
    } else {
      status.failure!!
    }
  }

  suspend fun requestLocationPermission(permission: PermissionRequest): Result {
    val validity = validateServiceStatus(permission)
    return if (validity.isValid) {
      Result.success(true)
    } else {
      validity.failure!!
    }
  }

  suspend fun lastKnownLocation(permission: Permission): Result {
    val validity = validateServiceStatus(permission)
    if (!validity.isValid) {
      return validity.failure!!
    }

    val location = try {
      lastLocation()
    } catch (e: Exception) {
      return Result.failure(Result.Error.Type.Runtime, message = e.message)
    }

    return if (location != null) {
      Result.success(arrayOf(LocationData.from(location)))
    } else {
      Result.failure(Result.Error.Type.LocationNotFound)
    }
  }


  // Updates API

  suspend fun addLocationUpdatesRequest(request: LocationUpdatesRequest) {
    val validity = validateServiceStatus(request.permission)
    if (!validity.isValid) {
      onLocationUpdatesResult(validity.failure!!)
      return
    }

    locationUpdatesRequests.add(request)
    updateLocationRequest()
  }

  fun removeLocationUpdatesRequest(id: Int) {
    locationUpdatesRequests.removeAll { it.id == id }
    updateLocationRequest()
  }

  fun registerLocationUpdatesCallback(callback: (Result) -> Unit) {
    check(locationUpdatesCallback == null) { "trying to register a 2nd location updates callback" }
    locationUpdatesCallback = callback
  }

  fun deregisterLocationUpdatesCallback() {
    check(locationUpdatesCallback != null) { "trying to deregister a non-existent location updates callback" }
    locationUpdatesCallback = null
  }


  // Lifecycle API

  fun resume() {
    if (!hasLocationRequest || !isPaused) {
      return
    }

    isPaused = false
    updateLocationRequest()
  }

  fun pause() {
    if (!hasLocationRequest || isPaused || hasInBackgroundLocationRequest) {
      return
    }

    isPaused = true
    updateLocationRequest()
    providerClient.removeLocationUpdates(locationCallback)
  }


  // Location updates logic

  private fun onLocationUpdatesResult(result: Result) {
    locationUpdatesCallback?.invoke(result)
  }

  private fun updateLocationRequest() {
    GlobalScope.launch(Dispatchers.Main) {
      if (locationUpdatesRequests.isEmpty()) {
        currentLocationRequest = null
        isPaused = false
        providerClient.removeLocationUpdates(locationCallback)
        return@launch
      }

      if (currentLocationRequest != null) {
        providerClient.removeLocationUpdates(locationCallback)
      }

      if (isPaused) {
        return@launch
      }

      val hasCurrentRequest = locationUpdatesRequests.any { it.strategy == LocationUpdatesRequest.Strategy.Current }
      if (hasCurrentRequest) {
        val lastLocationResult = lastLocationIfAvailable()
        if (lastLocationResult != null) {
          onLocationUpdatesResult(lastLocationResult)

          val hasOnlyCurrentRequest = locationUpdatesRequests.all { it.strategy == LocationUpdatesRequest.Strategy.Current }
          if (hasOnlyCurrentRequest) {
            return@launch
          }
        }
      }

      val locationRequest = LocationRequest.create()

      locationRequest.priority = locationUpdatesRequests.map { it.accuracy.androidValue }
          .sortedWith(Comparator { o1, o2 ->
            when (o1) {
              o2 -> 0
              LocationHelper.getBestPriority(o1, o2) -> 1
              else -> -1
            }
          })
          .first()

      locationUpdatesRequests.map { it.displacementFilter }
          .min()!!
          .takeIf { it > 0 }
          ?.let { locationRequest.smallestDisplacement = it }

      locationUpdatesRequests
          .filter { it.options.interval != null }
          .map { it.options.interval!! }
          .min()
          ?.let { locationRequest.interval = it }

      locationUpdatesRequests
          .filter { it.options.fastestInterval != null }
          .map { it.options.fastestInterval!! }
          .min()
          ?.let { locationRequest.fastestInterval = it }

      locationUpdatesRequests
          .filter { it.options.expirationTime != null }
          .map { it.options.expirationTime!! }
          .min()
          ?.let { locationRequest.expirationTime = it }

      locationUpdatesRequests
          .filter { it.options.expirationDuration != null }
          .map { it.options.expirationDuration!! }
          .min()
          ?.let { locationRequest.setExpirationDuration(it) }

      locationUpdatesRequests
          .filter { it.options.maxWaitTime != null }
          .map { it.options.maxWaitTime!! }
          .min()
          ?.let { locationRequest.maxWaitTime = it }

      if (locationUpdatesRequests.any { it.strategy == LocationUpdatesRequest.Strategy.Continuous }) {
        locationUpdatesRequests
            .filter { it.options.numUpdates != null }
            .map { it.options.numUpdates!! }
            .max()
            ?.let { locationRequest.numUpdates = it }
      } else {
        locationRequest.numUpdates = 1
      }

      currentLocationRequest = locationRequest

      if (!isPaused) {
        providerClient.requestLocationUpdates(currentLocationRequest!!, locationCallback, null)
      }
    }
  }

  private suspend fun lastLocationIfAvailable(): Result? {
    return try {
      val availability = locationAvailability()
      if (availability.isLocationAvailable) {
        val location = lastLocation()
        if (location != null) {
          Result.success(arrayOf(LocationData.from(location)))
        } else {
          null
        }
      } else {
        null
      }
    } catch (e: Exception) {
      Result.failure(Result.Error.Type.Runtime, message = e.message)
    }
//
//        if (result != null) {
//            onLocationUpdatesResult(result)
//        }
//
//        return result != null
  }


  // Service status

  private suspend fun validateServiceStatus(permission: Permission): ValidateServiceStatus {
    return validateServiceStatus(PermissionRequest(permission, openSettingsIfDenied = false))
  }

  private suspend fun validateServiceStatus(permission: PermissionRequest): ValidateServiceStatus {
    val status = currentServiceStatus(permission.value)
    if (status.isReady) return ValidateServiceStatus(true)

    return if (status.needsAuthorization) {
      if (requestPermission(permission.value)) {
        ValidateServiceStatus(true)
      } else {
        ValidateServiceStatus(false, Result.failure(Result.Error.Type.PermissionDenied))
      }
    } else if (status.failure!!.error!!.type == Result.Error.Type.PermissionDenied && permission.openSettingsIfDenied && tryShowSettings()) {
      val refreshedStatus = currentServiceStatus(permission.value)
      return if (refreshedStatus.isReady) {
        ValidateServiceStatus(true)
      } else {
        ValidateServiceStatus(false, refreshedStatus.failure)
      }
    } else {
      ValidateServiceStatus(false, status.failure!!)
    }
  }


  private fun currentServiceStatus(permission: Permission): ServiceStatus {
    val connectionResult = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
    if (connectionResult != ConnectionResult.SUCCESS) {
      return ServiceStatus(isReady = false, needsAuthorization = false,
          failure = Result.failure(Result.Error.Type.PlayServicesUnavailable,
              playServices = Result.Error.PlayServices.fromConnectionResult(connectionResult)))
    }

    if (!LocationHelper.isLocationEnabled(context)) {
      return ServiceStatus(isReady = false, needsAuthorization = false, failure = Result.failure(Result.Error.Type.ServiceDisabled))
    }

    if (!LocationHelper.isPermissionDeclared(context, permission)) {
      return ServiceStatus(isReady = false, needsAuthorization = false, failure = Result.failure(Result.Error.Type.Runtime, message = "Missing location permission in AndroidManifest.xml. You need to add one of ACCESS_FINE_LOCATION or ACCESS_COARSE_LOCATION. See readme for details.", fatal = true))
    }

    if (activity != null && LocationHelper.isPermissionDeclined(activity!!, permission)) {
      return ServiceStatus(isReady = false, needsAuthorization = false, failure = Result.failure(Result.Error.Type.PermissionDenied))
    }

    if (!LocationHelper.isPermissionGranted(context)) {
      return ServiceStatus(isReady = false, needsAuthorization = true, failure = Result.failure(Result.Error.Type.PermissionNotGranted))
    }

    return ServiceStatus(true)
  }


  // Permission

  private suspend fun tryShowSettings(): Boolean = suspendCoroutine { cont ->
    if (activity == null) {
      cont.resume(false)
      return@suspendCoroutine
    }

    permissionSettingsCallback = {
      cont.resume(true)
    }

    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.fromParts("package", context.packageName, null))
    activity!!.startActivityForResult(intent, GeolocationPlugin.Intents.LocationPermissionSettingsRequestId)
  }

  private suspend fun requestPermission(permission: Permission): Boolean = suspendCoroutine { cont ->
    if (activity == null) {
      cont.resume(false)
      return@suspendCoroutine
    }

    val callback = Callback<Unit, Unit>(
        success = { cont.resume(true) },
        failure = { cont.resume(false) }
    )
    permissionCallbacks.add(callback)

    ActivityCompat.requestPermissions(activity!!, arrayOf(permission.manifestValue), GeolocationPlugin.Intents.LocationPermissionRequestId)
  }

  private suspend fun requestEnablingLocation(): Boolean = suspendCoroutine { cont ->
    val callback = Callback<Unit, Unit>(
        success = { cont.resume(true) },
        failure = { cont.resume(false) }
    )

    val request = LocationRequest()
    request.priority = LocationRequest.PRIORITY_HIGH_ACCURACY

    LocationServices
        .getSettingsClient(context)
        .checkLocationSettings(
            LocationSettingsRequest
                .Builder()
                .addLocationRequest(request)
                .setAlwaysShow(true)
                .build()
        ).addOnSuccessListener {
          callback.success(Unit)
        }.addOnFailureListener { exception ->
          when (exception) {
            is ApiException -> when (exception.statusCode) {
              LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
                if (activity == null) {
                  callback.failure(Unit)
                  return@addOnFailureListener
                }

                try {
                  val resolvable = exception as ResolvableApiException
                  resolvable.startResolutionForResult(activity, GeolocationPlugin.Intents.EnableLocationSettingsRequestId)
                  locationSettingsCallbacks.add(callback)
                } catch (ignore: java.lang.Exception) {
                  callback.failure(Unit)
                }
              }

              else -> {
                callback.failure(Unit)
              }
            }

            else -> {
              callback.failure(Unit)
            }
          }
        }
  }

  // Location helpers

  private suspend fun locationAvailability(): LocationAvailability = suspendCoroutine { cont ->
    providerClient.locationAvailability
        .addOnSuccessListener { cont.resume(it) }
        .addOnFailureListener { cont.resumeWithException(it) }
  }

  private suspend fun lastLocation(): Location? = suspendCoroutine { cont ->
    providerClient.lastLocation
        .addOnSuccessListener { location: Location? -> cont.resume(location) }
        .addOnFailureListener { cont.resumeWithException(it) }
  }


  // Structures

  private class Callback<in T, in E>(val success: (T) -> Unit, val failure: (E) -> Unit)

  private class ServiceStatus(val isReady: Boolean,
                              val needsAuthorization: Boolean = false,
                              val failure: Result? = null)

  private class ValidateServiceStatus(val isValid: Boolean, val failure: Result? = null)
}
